### ISSUE TITLE
Redundant config syncs

### ISSUE THREAD

**Issue by munen** [200ok-ch/organice#736]

**Comment by munen** [comment]
Taken from https://github.com/200ok-ch/organice/pull/734#discussion_r740365312 and slightly rephrased:

The intended strategy interface can also use some tweaking. It seems like since there are separate `createFile` and `updateFile` functions that the GitLab implementation is "correct", because otherwise the strategy interface would have a single `createOrUpdateFile` function instead of distinct create vs update functions. I realize the other backends allow updating an existing file via `createFile()` but that seems like misuse of a leaky abstraction as opposed to intended behavior. 

In theory, it seems like the correct implementation of `debouncedPushConfigToSyncBackend` should be something like one of these two options (rough and untested, just for illustration):
```javascript
// Pre-emptively figure out whether create or update is the right call to make. This makes sense
// if there's an easy way to track this via local state, but I'm not sure how straightforward
// that would be. If the only way to determine if fileExists is by something like
// syncBackendClient.getFileContents(..) then I think the other example is better.
const debouncedPushConfigToSyncBackend = _.debounce(
  (syncBackendClient, contents) => {
    const fileExists = figureThisOut();
    let promise;
    if (fileExists) {
      promise = syncBackendClient.updateFile('/.organice-config.json', contents);
    } else {
      promise = syncBackendClient.createFile('/.organice-config.json', contents);
    }
    promise.catch((error) =>
      alert(`There was an error trying to push settings to your sync backend: ${error}`)
    );
  },
  1000,
  { maxWait: 3000 }
);

// Try updating first because the file is going to exist most of the time this is called, then
// fall back to create if it fails.
// This is likely a better idea than the other example.
const debouncedPushConfigToSyncBackend = _.debounce(
  (syncBackendClient, contents) => {
    syncBackendClient.updateFile('/.organice-config.json', contents).catch(() => {
      syncBackendClient
        .createFile('/.organice-config.json', contents)
        .catch((error) =>
          alert(`There was an error trying to push settings to your sync backend: ${error}`)
        );
    });
  },
  1000,
  { maxWait: 3000 }
);
```

Unfortunately the Google Drive backend breaks the strategy interface completely with the way it takes an extra 'root' argument, so that would have to be changed too, which would require a non-trivial amount of work I imagine. 

Everything Google Drive is unfortunate (of course I mean only in organice cough). We've been waiting for aeons and I've personally put a lot of effort into trying to get [Google to enable production mode for the Drive App](https://github.com/200ok-ch/organice/issues/127). So far without success. There's other issues with the Drive integration, too. That it is now also in the way of creating a more solid sync back-end abstraction only compounds the already huge issue around Google Drive.

So the next step is either a complete cleanup and getting the Drive app to work or just to remove the code. Any contributor who runs into this the next time is welcome to remove it. I'm not aware that any maintainer is using it. If somebody else wants to step up and fix it, they can still do it.

**Comment by chasecaleb** [comment]
Good summary from my comment. I agree with your analysis about Google Drive too. Unfortunately that seems about par for the course with Google when it comes to support.

Here's a couple ideas for determining Google Drive usage without client-side analytics:
1. Does Google Drive's OAuth integration offer any insight into client usage? GitLab shows a client count for example, although that's significantly over-inflated because it goes off the number of initial (i.e. non-refreshed) tokens issued, not the number of accounts using. For instance my GitLab organice test application shows 14 clients due to signing in and out while developing, despite only using it on one computer/browser with two accounts.
1. We could add a notification (e.g. `alert()`, modal, or something at the top of the page) within the app shown only to Google Drive users. It could ask them to comment here, or maybe better yet give them a link to a survey type website with a simple "Are you using Google Drive? yes/no" form that we could count the number of submissions on.

Not sure if that's worth dealing with though. Just pointing them out while it's fresh on my mind.

**Comment by munen** [comment]
> Good summary from my comment. I agree with your analysis about Google Drive too. Unfortunately that seems about par for the course with Google when it comes to support.

:+1: 

> Does Google Drive's OAuth integration offer any insight into client usage?

It does. And the metrics would be 'more accurate' than the ones from GitLab But it doesn't help a whole lot, because the non-production OAuth app only allows 200 (iirc) users which we have exceeded a long time ago. That's why the sign in screen shows: 
![image](https://user-images.githubusercontent.com/505721/139713503-fee7bf79-cbdd-447d-851e-7bcc2fad7ab4.png)

There's no way to tell how many people did that since we have no analytics whatsoever. For these kinds of decisions, it would be very interesting, though. I'm personally also curious to know how many people have their own instance overall. But without analytics, there's no way of knowing(;

> Not sure if that's worth dealing with though. Just pointing them out while it's fresh on my mind.

You're ideas are very nice and they are much appreciated. However, the whole Google Drive thing is a rabbit hole that I have entered a few too many times. Quite a lot of users spoke up on the issue of not being able to login to Google Drive on the reference instance, but nobody took up the slack with refactoring code. I'm neither a Google Drive user nor a big fan of Google, but I did put in a couple days of my life into this. To be honest, I feel that this has been fair enough. The potential Google Drive users have been given fair warning and I've called for support a few times as well.

I mean it when I say that the next time Google Drive is in the way, it'll just get cut. organice is a FLOSS application. If there's any Google Drive user who want to continue using the current version of organice after the cut, they can make a PR based off the last version before the deletion. I'll be very happy to support them in the integration as I have been with your PR. However, I'm not going to put any additional effort into fixing a fraction of the issues that the current Google Drive integration has for a hypothetical user.

I realize these words might sound harsh. They are not intended as such. It's late and I should get some dinner. So I write quickly and just put down a stream of tought^^

**Comment by chasecaleb** [comment]
> I realize these words might sound harsh.

Not at all! I wholeheartedly agree. That's the nice thing about a FLOSS project compared to a job.

---

So **to anyone who cares about ongoing Google Drive functionality**: please help out with this soon, or at the bare minimum comment on this issue if you're planning to contribute changes soon. Otherwise the first thing that I do if/when I work on this issue will be to remove Google Drive from organice.

I don't plan on working on this issue anytime soon unless it gets in my way because I have other priorities, but I might within the next few months.

**Comment by munen** [comment]
Sweet!

Just to put all relevant information together in one place. These are the big issues with Google Drive at the moment:

1. https://github.com/200ok-ch/organice/issues/127
2. https://github.com/200ok-ch/organice/issues/109
3. https://github.com/200ok-ch/organice/issues/107

Any user (not just Google Drive users) will already benefit when #109 is dealt with. It will speed organice up and cause less traffic. Not to speak of that we don't have to put our trust in that JS file that we download from Google anymore.
